Assets/MixedRealityToolkit/MSBuild/Scripts/CSProjectInfo.cs

```diff
@@ -0,0 +1,363 @@
+
+        /// <summary>
+        /// The assembly that backs this object.
+        /// </summary>
+        public Assembly Assembly { get; }
+
+        /// <summary>
+        /// The associated assembly definition info if available.
+        /// </summary>
+        public AssemblyDefinitionInfo AssemblyDefinitionInfo { get; }
+
+        /// <summary>
+        /// The type of the project.
+        /// </summary>
+        public ProjectType ProjectType { get; }
+
+        /// <summary>
+        /// Creates a new instance of the CSProject info.
+        /// </summary>
+        /// <param name="availablePlatforms">A list of platforms available to MSBuild.</param>
+        /// <param name="guid">The unique Guid of this reference item.</param>
+        /// <param name="assemblyDefinitionInfo">The associated assembly definition info.</param>
+        /// <param name="assembly">The Unity assembly object associated with this csproj.</param>
+        /// <param name="baseOutputPath">The output path where everything will be outputted.</param>
+        internal CSProjectInfo(IEnumerable<CompilationPlatformInfo> availablePlatforms, Guid guid, AssemblyDefinitionInfo assemblyDefinitionInfo, Assembly assembly, string baseOutputPath)
+            : base(availablePlatforms, guid, new Uri(Path.Combine(baseOutputPath, $"{assembly.name}.csproj")), assembly.name)
+        {
+            AssemblyDefinitionInfo = assemblyDefinitionInfo;
+            Assembly = assembly;
+
+            ProjectType = GetProjectType(assemblyDefinitionInfo, assembly);
+
+            InEditorPlatforms = GetCompilationPlatforms(true);
+            PlayerPlatforms = GetCompilationPlatforms(false);
+
+            if (InEditorPlatforms.Count == 0 && PlayerPlatforms.Count == 0)
+            {
+                Debug.LogError($"The assembly project '{Name}' doesn't contain any supported in-editor or player platform targets.");
+            }
+        }
+
+        private ProjectType GetProjectType(AssemblyDefinitionInfo assemblyDefinitionInfo, Assembly assembly)
+        {
+            if (assemblyDefinitionInfo != null)
+            {
+                return assemblyDefinitionInfo.EditorPlatformSupported && !assemblyDefinitionInfo.NonEditorPlatformSupported ? ProjectType.EditorAsmDef : ProjectType.AsmDef;
+            }
+
+            switch (assembly.name)
+            {
+                case "Assembly-CSharp":
+                case "Assembly-CSharp-firstpass":
+                    return ProjectType.PredefinedAssembly;
+                case "Assembly-CSharp-editor":
+                case "Assembly-CSharp-firstpass-editor":
+                    return ProjectType.PredefinedEditorAssembly;
+                default:
+                    Debug.LogError($"Predefined assembly '{assembly.name}' was not recognized, this generally means it should be added to the switch statement in CSProjectInfo:GetProjectType. Treating is as a PredefinedAssembly instead of PredefinedEditorAssembly.");
```

Comment:
>Debug.LogError($"Predefined assembly '{assembly.name}' was not recognized, this generally means it should be added to the switch statement in CSProjectInfo:GetProjectType. Treating is as a PredefinedAssembly instead of PredefinedEditorAssembly."); [](start = 20, length = 247)

Another question about how this surfaces in the build - it would be useful if we could throw/abort the build entirely (if we aren't going to invest in build logging changes to grab from STDERR or parse errors). If the plan, however, is to update the build steps to look for error logs (of this specific type) then this is fine - my main concern is that usually we haven't done this, and these sorts of things can fail silently (because they only get exercised in batch mode and on CI usually we don't look at the logged errors, only the overall failure/success).

The downside to throwing, of course, is that we'll only see a single error (and we would have to re-run things possibly multiple times to get all errors, so really there are pros and cons to both ways)